### PR TITLE
Use 2-chain for finality

### PR DESCRIPTION
**Pull request by linh2931**
* Update `finality_core`  to use two-chain for finality.
* Update the proofs in `get_new_block_numbers()`.
* Update all relevant tests.

Resolves https://github.com/AntelopeIO/spring/issues/384

### CODE REVIEW

**Review by heifner — APPROVED** [review]

**Review by systemzax — APPROVED** [review]

File: unittests/finality_tests.cpp [hunk 1]
```diff
@@ -746,15 +746,15 @@ BOOST_FIXTURE_TEST_CASE(finality_skip, finality_test_cluster<4>) { try {
    // `generation == 4`. The other policies must have been overwritten since they all
    // became `pending` at the same block.
    //
-   // we need another 3-chain to make that block final.
+   // we need another 2-chain to make that block final.
    // -------------------------------------------------------------------------------
-   for (size_t i=0; i<3; ++i) {
+   for (size_t i=0; i<(num_chains_to_final - 1); ++i) {
```
Comment (greg7mdp) [inline unittests/finality_tests.cpp:751]: If we need a 2-chain, why `num_chains_to_final - 1`.  Also the comment at line 733 above is incorrect, we won't have three pending policies at the same time.

File: unittests/finalizer_update_tests.cpp [hunk 2]
```diff
@@ -52,10 +52,10 @@ BOOST_AUTO_TEST_CASE(savanna_set_finalizer_single_test) { try {
    t.produce_block();
    t.check_head_finalizer_policy(1u, pubkeys0); // new policy should only be active until after two 3-chains
 
-   t.produce_blocks(3);
+   t.produce_blocks(num_chains_to_final);
    t.check_head_finalizer_policy(1u, pubkeys0); // one 3-chain - new policy still should not be active
 
-   t.produce_blocks(2);
+   t.produce_blocks(num_chains_to_final-1);
    t.check_head_finalizer_policy(1u, pubkeys0); // one 3-chain + 2 blocks - new policy still should not be active
```
Comment (greg7mdp) [inline unittests/finalizer_update_tests.cpp:59]: comment is wrong, it should be `+ 1 block`

File: unittests/finalizer_update_tests.cpp [hunk 3]
```diff
@@ -101,8 +101,8 @@ BOOST_AUTO_TEST_CASE(savanna_set_finalizer_multiple_test) { try {
    auto pubkeys2 = fin_keys.set_finalizer_policy(2u).pubkeys;
    t.produce_block();
    t.check_head_finalizer_policy(1u, pubkeys0); // new policy should only be active until after two 3-chains
```
Comment (greg7mdp) [inline unittests/finalizer_update_tests.cpp:103]: many comments in this file still mention `3-chains`. Some in other files as well.

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: Switch to 2-chain for finality.
Note:end

File: unittests/finalizer_update_tests.cpp [hunk 4]
```diff
@@ -120,42 +120,42 @@ BOOST_AUTO_TEST_CASE(savanna_set_finalizer_multiple_test) { try {
...
    b = t.produce_block(); // pending: pubkeys5, proposed: pubkeysr6, pubkeys7, pubkeys8
    verify_block_finality_policy_diff(b, 8, pubkeys8.back());
    auto pubkeys9 = fin_keys.set_finalizer_policy(9u).pubkeys;
-   t.check_head_finalizer_policy(3u, pubkeys3);
+   t.check_head_finalizer_policy(4u, pubkeys4);
    b = t.produce_block(); // pending: pubkeys5, proposed: pubkeysr6, pubkeys7, pubkeys8, pubkeys9
    verify_block_finality_policy_diff(b, 9, pubkeys9.back());
    auto pubkeys10 = fin_keys.set_finalizer_policy(10u).pubkeys;
-   t.check_head_finalizer_policy(3u, pubkeys3); // 7 blocks after pubkeys4, pubkeys4 should still be active
+   t.check_head_finalizer_policy(5u, pubkeys5); // 7 blocks after pubkeys4, pubkeys4 should still be active
```
Comment (greg7mdp) [inline unittests/finalizer_update_tests.cpp:140]: comment is clearly incorrect.

File: unittests/finalizer_update_tests.cpp (hunk 2 above)
Comment (linh2931) [reply unittests/finalizer_update_tests.cpp:59]: Fixed it by https://github.com/AntelopeIO/spring/pull/389/commits/c1feb767a7ede694db9dae81d0af3ccf57721936

Thanks. I am updating comments in other files.

File: unittests/finality_tests.cpp (hunk 1 above)
Comment (linh2931) [reply unittests/finality_tests.cpp]: Fixed it by https://github.com/AntelopeIO/spring/pull/389/commits/c1feb767a7ede694db9dae81d0af3ccf57721936

Thanks.

File: unittests/finalizer_update_tests.cpp (hunk 3 above)
Comment (linh2931) [reply unittests/finalizer_update_tests.cpp:103]: To be fixed by https://github.com/AntelopeIO/spring/issues/398

File: unittests/finalizer_update_tests.cpp (hunk 4 above)
Comment (linh2931) [reply unittests/finalizer_update_tests.cpp:140]: To be fixed by https://github.com/AntelopeIO/spring/issues/398

**Pull request merged**